File: population_growth/growth.go
package main

import (
	"flag"
	"fmt"
	"math"
	"time"
)

// use math to solve the problem
// helper function, where all args are converted to floats
// Proof of this solution can be found in this repo under GrowthOfAPopulation directory (this solution directory)
func nbYearFloat(p0 float64, percent float64, aug float64, p float64) int {
	// Handle case, when growth rate is 1 (meaning the geom series that is related with aug does not converge)
	if percent == 0 {
		return int((p - p0) / aug)
	}
	r := 1 + percent/100
	c := (aug + p*r - p) / (aug + p0*r - p0)
	return int(math.Ceil(math.Log(c) / math.Log(r)))
}

// use math to solve the problem
func NbYear(p0 int, percent float64, aug int, p int) int {
	return nbYearFloat(float64(p0), percent, float64(aug), float64(p))
}

// using iterative method to solve the problem
func NbYearForce(p0 int, percent float64, aug int, p int) int {
	var pIt float64 = float64(p0)
	var counter int
	for pIt < float64(p) {
		counter += 1
		pIt = (1+percent/100)*pIt + float64(aug)
	}
	return counter
}

func main() {
	p0 := flag.Int("p0", 0, "P_0")
	p := flag.Float64("p", 0, "percent")
	aug := flag.Int("aug", 0, "augment")
	pn := flag.Int("pn", 0, "P_N")
	flag.Parse()
	if (*p0) == 0 && (*p) == 0 && (*aug) == 0 && (*pn) == 0 {
		(*p0) = 100
		(*p) = 0.001
		(*aug) = 10
		(*pn) = 1000000000000000000
	}
	fmt.Printf("P_0: %v\ngrowth: %v\nnew inhab.: %v\n", *p0, *p, *aug)
	i := 1000
	fmt.Printf("%-20s%-20s%-20s%-20s%-20s%-20s\n", "P_N", "Calculation time", "Iteration time", "Calculated value", "Iterated value", "Speedup")
	for i < (*pn) {
		startCalc := time.Now()
		calculated := NbYear(*p0, *p, *aug, i)
		endCalc := time.Since(startCalc)
		startIter := time.Now()
		iterated := NbYearForce(*p0, *p, *aug, i)
		endIter := time.Since(startIter)
		fmt.Printf("%-20v%-20v%-20s%-20v%-20v%-20v\n", i, endCalc, endIter, calculated, iterated, int64(endIter/endCalc))
		i *= 10
	}
}

File: population_growth/growth.py
import time
import math

def nb_year(p0,percent,aug,p):
    if percent == 0 :
        return (p-p0)/aug
    r = 1+percent/100
    c = (aug+p*r-p)/(aug+p0*r-p0)
    return math.ceil(math.log(c)/math.log(r))

def nb_year_iter(p0,percent,aug,p):
    count = 0
    while p0 < p:
        count += 1
        p0 = (1+percent/100)*p0 + aug
    return count

i = 1000
p0 = 100
p = 0.001
aug = 10
pn = 100_000_000_000_000_000
print(f"P_0: {p0}\ngrowth: {p}\nnew inhab.: {aug}")
print(f'{"P_N":<20}{"Calculation time":<20}{"Iteration time":<20}{"Calculation value":<20}{"Iteration value":<20}{"Speedup":<20}')
while i < pn:
    start_calc=time.monotonic()
    calculated = nb_year(p0,p,aug,i)
    end_calc=time.monotonic()
    iterated = start_iter=time.monotonic()
    iterated = nb_year_iter(p0,p,aug,i)
    end_iter=time.monotonic()
    calc_time = end_calc-start_calc
    iter_time = end_iter-start_iter
    print(f'{i:<20}{round(calc_time,10):<20.2e}{round(iter_time,10):<20.2e}{calculated:<20}{iterated:<20}{round(iter_time/calc_time,2):<20}')
    i *= 10